Objective-C/CBLReplicatorConfiguration.m

```diff
@@ -189,31 +211,95 @@ - (void) setEnableAutoPurge: (BOOL)enableAutoPurge {
+                    format: @"Attempt to add an invalid collection"];
+    }
     
-    // TODO: Add implementation
-
+    if (_collections.count > 0) {
+        if (_collections[0].db != collection.db) {
+            [NSException raise: NSInvalidArgumentException
+                        format: @"Attempt to add collection from different databases"];
+        }
+    }
+    
+    CBLCollection* defaultCollection = [_database defaultCollectionOrThrow];
+    if (collection == defaultCollection)
+        [self checkAndUpdateConfig: config];
+    
+    CBLCollectionConfiguration* colConfig = nil;
+    if (config)
+        colConfig = [[CBLCollectionConfiguration alloc] initWithConfig: config];
+    else
+        colConfig = [[CBLCollectionConfiguration alloc] init];
+    
+    [_collectionConfigs setObject: colConfig forKey: collection];
+    
+    _collections = _collectionConfigs.allKeys;
 }
 
-- (void) addCollections: (NSArray*)collections
+- (void) addCollections: (NSArray<CBLCollection*>*)collections
                  config: (nullable CBLCollectionConfiguration*)config {
-
-    // TODO: Add implementation
+    Assert(collections.count > 0);
     
+    CBLDatabase* db = collections[0].db;
+    for (CBLCollection* col in collections) {
+        if (!col.isValid) {
+            [NSException raise: NSInvalidArgumentException
+                        format: @"Attempt to add an invalid collection"];
+        }
+        
+        if (db != col.db) {
+            [NSException raise: NSInvalidArgumentException
+                        format: @"Attempt to add collections from different databases"];
+        }
+    }
+    
+    CBLCollection* defaultCollection = [_database defaultCollectionOrThrow];
+    for (CBLCollection* col in collections) {
+        if (col == defaultCollection)
+            [self checkAndUpdateConfig: config];
+        
+        CBLCollectionConfiguration* colConfig = nil;
+        if (config)
+            colConfig = [[CBLCollectionConfiguration alloc] initWithConfig: config];
+        else
+            colConfig = [[CBLCollectionConfiguration alloc] init];
+        
+        [_collectionConfigs setObject: colConfig forKey: col];
+    }
+    _collections = _collectionConfigs.allKeys;
+}
+
+- (void) checkAndUpdateConfig: (CBLCollectionConfiguration*)config {
+    if (config.conflictResolver != _conflictResolver)
+        _conflictResolver = config.conflictResolver;
+    if (config.channels != _channels)
+        _channels = config.channels;
+    if (config.documentIDs != _documentIDs)
+        _documentIDs = config.documentIDs;
+    if (config.pushFilter != _pushFilter)
+        _pushFilter = config.pushFilter;
+    if (config.pullFilter != _pullFilter)
+        _pullFilter = config.pullFilter;
 }
 
 - (void) removeCollection:(CBLCollection *)collection {
-    
-    // TODO: Add implementation
-    
+    [_collectionConfigs removeObjectForKey: collection];
+    _collections = _collectionConfigs.allKeys;
```

Comment:
Same as above, add a getter for _collections instead of updating it every operation.

Reply:
Need to reset the _database if no collections.